[CodeGen][Spill2Reg] Initial patch

vporpo opened the pull request:
This is the first commit for the Spill2Reg optimization pass. The goal of this pass is to selectively replace spills to the stack with spills to vector registers. This can help remove back-end stalls in x86.

Old code review: https://reviews.llvm.org/D118298

RFC:
https://lists.llvm.org/pipermail/llvm-dev/2022-January/154782.html https://discourse.llvm.org/t/rfc-spill2reg-selectively-replace-spills-to-stack-with-spills-to-vector-registers/59630

Commit:
[CodeGen][Spill2Reg] Initial patch

This is the first commit for the Spill2Reg optimization pass.
The goal of this pass is to selectively replace spills to the stack with
spills to other registers. This can help remove back-end stalls in x86.

Old code review: https://reviews.llvm.org/D118298

RFC:
https://lists.llvm.org/pipermail/llvm-dev/2022-January/154782.html
https://discourse.llvm.org/t/rfc-spill2reg-selectively-replace-spills-to-stack-with-spills-to-vector-registers/59630

Commit:
[Spill2Reg] This patch adds spill/reload collection

Walk through the code looking for spills and reloads and group them
per stack slot.

Original review: https://reviews.llvm.org/D118299

Commit:
[Spill2Reg] Code generation part 1

This patch adds the main structure of the code generation phase of Spill2Reg.
Iterate through the spills/reloads collected earlier and generate the
new instructions.

Original review: https://reviews.llvm.org/D118300

Commit:
[Spill2Reg] Code generation part 2

Spill2Reg can now emit spill and reload instructions.
This will not generate correct code, as it does not keep track of live regs.

Original review: https://reviews.llvm.org/D118302

Commit:
[Spill2Reg] Add live register tracking

This patch implements tracking of live registers. This is used to look for free
vector registers. It works by walking up the CFG from the reloads all the way
to the spills, accumulating the register units being used.
This implementation caches the live register units used by each MBB for faster
compilation time.

Note: Live register tracking relies on MBB liveins/outs being maintained
correctly, which is implemented in a follow-up patch. So this patch will still
not generate correct code for all but some simple cases.

Original review: https://reviews.llvm.org/D118303

Commit:
[Spill2Reg] MBB live-ins are now being updated

This patch implements updates for the MBB live-ins due to the newly introduced
instructions emitted by spill2reg.
This is required for correct tracking of live register usage.

Original review: https://reviews.llvm.org/D118304

Commit:
[Spill2Reg] Adds code generation for 8/16bit spill/reloads in x86

This patch adds support for 8/16 bit values in x86.

Original review: https://reviews.llvm.org/D118305

Commit:
[Spill2Reg] Use AVX opcodes when available

This patch updates the vector spill/reload instructions to use the AVX
opcodes by default if the targets supports it. This can be turned off
with the -spill2reg-no-avx flag.

Original review: https://reviews.llvm.org/D118951

williamweixiao commented:
There are some DFS on the CFG and do you have data about the impact to compilation speed?

vporpo reviewed (commented):
> There are some DFS on the CFG and do you have data about the impact to compilation speed?

It's been a while since I tested this pass, but as far as I remember it wasn't too bad. The traversals done in the code generation phase of the pass are not too frequent as they only happen when we need to spill to vector registers, which shouldn't be too often.

williamweixiao commented on llvm/lib/Target/X86/X86InstrInfo.cpp:
```diff
@@ -10893,5 +10947,208 @@ void X86InstrInfo::getFrameIndexOperands(SmallVectorImpl<MachineOperand> &Ops,
+  if (X86::VK16RegClass.contains(Reg))
+    return false;
+
+  switch (unsigned Bits = TRI->getRegSizeInBits(Reg, *MRI)) {
```
"Bits" is unused variable.

vporpo replied:
Removed.

vporpo commented on llvm/lib/CodeGen/Spill2Reg.cpp:
```diff
@@ -0,0 +1,548 @@
+    LiveRegUnits LRU(*TRI);
+    calculateLiveRegs(Entry, LRU);
+
+    // Look for a physical register that in LRU.
```
Fixed.

vporpo commented on llvm/lib/CodeGen/Spill2Reg.cpp:
```diff
@@ -0,0 +1,548 @@
+        // compilation time.
+        for (auto &MID : Entry.Reloads)
+          if (MID.MI->getParent() == &MBB)
+            MID.IsLiveIn = false;
```
Yes, this looks wrong. I removed this code.

vporpo commented on llvm/lib/Target/X86/X86InstrInfo.cpp:
```diff
@@ -10893,5 +10947,208 @@ void X86InstrInfo::getFrameIndexOperands(SmallVectorImpl<MachineOperand> &Ops,
+      const unsigned MinVecBits =
+          TRI->getRegSizeInBits(*TRI->getRegClass(X86::VR128RegClassID));
+      if (MF->getFrameInfo().getObjectSize(MO.getIndex()) >= MinVecBits)
+        return true;
```
I don't have such an instruction in mind and I can't recall if I did back when I wrote this. It's probably a conservative check. I added a TODO to check if this is needed.

RKSimon reviewed (commented):
I'd like to ensure this patch doesn't get too focussed on just working for gpr->vector spills. In my experience those profitable cases are pretty rare. What has been more useful has been cases such as storing scalar f32/f64 in the upper elements of xmm registers, or even using ymm upper halfs to store xmm vector data, and to a lesser extent storing a i32 in the upper 32-bits of a i64 gpr.

vporpo commented on llvm/include/llvm/CodeGen/TargetInstrInfo.h:
```diff
@@ -294,6 +294,11 @@ class TargetInstrInfo : public MCInstrInfo {
 
+  virtual const MachineOperand *isLoadFromStackSlotMO(const MachineInstr &MI,
+                                                      int &FrameIndex) const {
+    llvm_unreachable("target did not implement");
```
The reasoning is that if I was implementing this for a new target I would prefer getting a crash telling me that I  should override this function, rather than getting it to silently skip spill2reg because some functions are not overridden. Wdyt?

vporpo commented on llvm/include/llvm/CodeGen/TargetInstrInfo.h:
```diff
@@ -2284,6 +2294,52 @@ class TargetInstrInfo : public MCInstrInfo {
+  }
+
+  virtual const TargetRegisterClass *
+  getVectorRegisterClassForSpill2Reg(const TargetRegisterInfo *TRI,
```
Makes sense, done. 

There are a couple more of these, like `spill2RegExtractFromVectorReg()`. I am thinking of using the term `Host` instead of `Vector` to describe the register used by spill2reg, what do you think?

vporpo commented on llvm/include/llvm/CodeGen/TargetInstrInfo.h:
```diff
@@ -2284,6 +2294,52 @@ class TargetInstrInfo : public MCInstrInfo {
+                                  const MachineRegisterInfo *MRI) const {
+    llvm_unreachable(
+        "Target didn't implement TargetInstrInfo::isLegalToSpill2Reg!");
+  }
```
I think most of them have to do with Instructions rather than registers, with the exception of `isLegalToSpill2Reg(), and `getVectorRegisterClassForSpill2Reg()`. I will move these to TargetRegisterInfo.
- `isStoreToStackSlotMO()` inspects an instruction
- `targetSupportsSpill2Reg()` could be placed in either file as it does not check an instruction or register
- `isSpill2RegProfitable()` checks the instruction sequence
- `spill2RegInsertToVectoReg()` emits instructions
- `spill2RegExtractFromVectorReg()` emits instructions

vporpo commented:
>> I'd like to ensure this patch doesn't get too focussed on just working for gpr->vector spills. In my experience those profitable cases are pretty rare. What has been more useful has been cases such as storing scalar f32/f64 in the upper elements of xmm registers, or even using ymm upper halfs to store xmm vector data, and to a lesser extent storing a i32 in the upper 32-bits of a i64 gpr.

> Yes, we also observed some cases in which spilling float value into GPR can help performance (suppose GPR register pressure is low meanwhile).

I think that the structure of the pass is already fairly agnostic to the variant of spill2reg (like GPR->lower vector, GPR->upper vector, GPR->upper GPR, F32/64->upper vector). The candidates are filtered by `TRI` callbacks, like `isLegalToSpillToReg()` and code generation is done with `TII` callbacks: `spill2RegInsertToVectorReg()` and `spill2RegExtractFromVectoReg()`. All of these can be update to work with different spill2reg variants.

Once we add support for more than one spill2reg variant, then during the collection phase we would need to determine the variant and set it in `StackSlotDataEntry` using an enum. This is going to be used later in the pass (in `generateCode()`) to generate the corresponding code.

vporpo commented:
Changed TII function names `spill2RegInsertToVectorReg()` to `spill2RegInsertToS2RReg()` and `spill2RegExtractFromVectorReg()` to `spill2RegExtractFromS2RReg()`. Rebased.

vporpo commented:
Shall we start focusing on one patch at a time for the code reviews? @williamweixiao @RKSimon any comments on the first patch?

williamweixiao commented on llvm/lib/CodeGen/Spill2Reg.cpp:
```diff
@@ -0,0 +1,531 @@
+      /// already walking through the code there. Otherwise we would need to
+      /// walk throught the code again in `updateLiveIns()` just to check for
+      /// other spills in the block, which would waste compilation time.
+      bool IsLiveIn = true;
```
It seems that "IsLiveIn" is always "true". Do we really need it?

vporpo replied:
I think a better solution is to set the `IsLiveIn` flag whenever we visit a reload in `collectSpillsAndReloads()`. I have updated the code.

vporpo commented on llvm/lib/CodeGen/Spill2Reg.cpp:
```diff
@@ -0,0 +1,534 @@
+    MachineBasicBlock *MBB = Reload->getParent();
+    bool IsSpillBlock = SpillMBBs.count(MBB);
+    // Add all MBB's live-outs.
+    LRU.addLiveOuts(*MBB);
```
I think you are right:

`AccumulateLRUUntilSpillFn()` is called in two cases:
1. when we calculate live regs from reload until a spill (after the call to `GetReloadLRU()`), in which case we have already calculated the live-outs at the point of the reload using `GetReloadLRU()`, starting from the live-outs and stepping backwards until the reload.
2. when we are looking for the spills, in which case we start from the bottom of the BB and walk up until the spill (called inside `AccumulateLRUFn()`). In this case we need to initialize the set  with the live-outs at MBB. 

So I think the issue is that we call `LRU.addLiveOuts()` inside `AccumulateLRUUntilSpillFn()`, instead of calling it only in `AccumulateLURFn()` just before we call `AccumulateLRUUntilSpillFn()`.

I have added some comments to make the code more readable.

vporpo commented:
Thank you for the comments @williamweixiao . Do you think we should start focusing on the individual patches of this PR one by one? There are 8 patches in the chain at this point which makes it hard to review and to maintain.

RKSimon commented:
Have you tried graphite to maintain the patch series?

vporpo commented:
> Have you tried graphite to maintain the patch series?

Thanks for the suggestion, I will give it a try. My main concern is testing: When the patch series is off-tree then we can only do so much testing on each change we are making, which can lead to regressions that get unnoticed. Given that there is enough interest in this project, my suggestion is to start working on this in-tree.
